Context Dependent Piecewise Linear Constraint

AleksandarZeljic opened the pull request:
This PR introduces an implementation of PiecewiseLinearConstraint using
context-dependent (CD) data structures from CVC4, which lazily and automatically
backtrack in sync with the central context object. These data structures
are delicate and require that data stored is safe to memcopy around. 
 
Members are initialized using the initializeCDOs( Context context ) method.
Subclasses need to call initializeDuplicatesCDOs() method in the
duplicateConstraint() method to avoid sharing context-dependent members.
CDOs need not be initialized for pre-processing purposes.

The class uses CDOs to record:
   * _cdConstraintActive
   * _cdPhaseStatus
   * _cdInfeasibleCases
 
PhaseStatus enumeration is used to keep track of current phase of the
object, as well as to record previously explored phases. This replaces
moving PiecewiseLinearCaseSplits around.
 
ContextDependentPiecewiseLinearConstraint stores locally which phases have
been explored so far using _cdInfeasibleCases. To communicate with the
search it implements two methods:
   * markInfeasible( PhaseStatus caseId ) - marks explored phase
   * nextFeasibleCase() - obtains next unexplored case if one exists
 
These methods rely on subclass properly handling:
   * _numCases field, which denotes the number of piecewise linear
      segments the constraint has, initialized by the constructor.
   * getAllCases() virtual method returns all possible phases represented using
     PhaseStatus enumeration. The order of returned cases will determine the
     search order, so this method should implement any search heuristics.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+class String;
+
+class ContextDependentPiecewiseLinearConstraint
+   : public virtual PiecewiseLinearConstraint
```
is virtual inheritance needed here?

AleksandarZeljic replied:
No, especially since the goal is to merge the two classes eventually. Fixed.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+     */
+    virtual PiecewiseLinearCaseSplit getCaseSplit( PhaseStatus caseId ) const = 0;
+
+
```
remove empty line

AleksandarZeljic replied:
Fixed.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+    /*
+      If the constraint's phase has been fixed, get the (valid) case split.
+    */
+    virtual PiecewiseLinearCaseSplit getImpliedCaseSplit() const = 0;
```
As a side note, we said we wanted to change this name, right? Maybe getImpliedCase, to be consistent with getAllCases()?

AleksandarZeljic replied:
So I would keep the name as is because there is a distinction in the return type - get*Case(s) all return PhaseStatus enumeration in some form, while this one actually returns an actual PiecewiseLinearCaseSplit. In the long run this should be the only method to return a PiecewiseLinearCaseSplit.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+    /*
+      Get the context object - debugging purposes only
+    */
+    CVC4::context::Context *getContext() const
```
Can we make the returned object const? i.e., const Context *
Same for the other debugging functions that return pointers

AleksandarZeljic replied:
Fixed.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+      Worst case complexity O(n^2)
+      This method should be overloaded for MAX and DISJUNCTION constraints.
+     */
+    PhaseStatus nextFeasibleCase();
```
If the method should be overloaded, shouldn't it be declared virtual?

AleksandarZeljic replied:
Fixed.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+    /*
+       Returns number of cases not yet marked as infeasible.
+     */
+    unsigned numFeasibleCases()
```
make function const?

AleksandarZeljic replied:
Fixed

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+    /*
+       Returns true if there is only one feasible case remaining.
+     */
+    bool isImplication()
```
Make function const?
Also, we usually do == 1 ad not 1 ==.

AleksandarZeljic replied:
Fixed, changed style. The idea is to avoid accidental assignment, but I understand that with our style and reviews this should be unneccessary.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.cpp:
```diff
@@ -0,0 +1,189 @@
+void ContextDependentPiecewiseLinearConstraint::registerBoundManager(
+    BoundManager *boundManager )
+{
+    ASSERT( nullptr == _boundManager );
```
we usually compare with the context on the right hand side

AleksandarZeljic replied:
Fixed

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.cpp:
```diff
@@ -0,0 +1,189 @@
+{
+    ASSERT( nullptr != _context );
+    ASSERT( nullptr == _cdConstraintActive );
+    _cdConstraintActive = new ( true ) CVC4::context::CDO<bool>( _context, true );
```
shouldn't we initialize to _constraintActive?

AleksandarZeljic replied:
Fixed

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.h:
```diff
@@ -0,0 +1,246 @@
+
+    /*
+      Returns phaseStatus representing next feasible case.
+      Returns PHASE_NOT_FIXED if no feasible case exists.
```
Can we perhaps add a dedicated sentinel? CONSTRAINT_INFEASIBLE or something?

AleksandarZeljic replied:
I have not done that because it would make dealing with Max and Disjunction trickier, but I will see what I can do.

AleksandarZeljic replied:
I was overthinking things, I added a high-valued entry to PhaseStatus enumeration, and updated this method accordingly.

AleksandarZeljic commented:
@guykatzz I think I've addressed all the concerns you raised, assuming the check passes, you can take another look. And if everything is fine I will rebase the branch onto master for a merge. Wasn't sure if rebase and force push would lose the review discussion or not.

guykatzz commented on src/engine/ContextDependentPiecewiseLinearConstraint.cpp:
```diff
@@ -0,0 +1,189 @@
+    ASSERT( _context != nullptr );
+    ASSERT( _cdPhaseStatus == nullptr );
+    _cdPhaseStatus =
+        new ( true ) CVC4::context::CDO<PhaseStatus>( _context, PHASE_NOT_FIXED );
```
initialize to _phaseStatus?

guykatzz reviewed (approved):
LGTM (but see my one comment)
